File: scenarios/ci/incremental.lua
require("pg_simple")

function main(argv)
	if fs.is_directory("backups") then
		warning("Deleting old stormweaver run backups")
		fs.delete_directory("backups")
	end

	simple_pg_single(argv, function(args)
		pgm:basebackup(1, "-D", "backups/backup_0", "-U", "stormweaver", "-c", "fast")

		for workloadIdx = 1, tonumber(args["repeat"]) do
			t1:run()
			t1:wait_completion()
			pgm:get(1, 10)

			prevBackupDir = "backups/backup_" .. tostring(workloadIdx - 1)
			backupDir = "backups/backup_" .. tostring(workloadIdx)

			pgm:basebackup(
				1,
				"-D",
				backupDir,
				"-U",
				"stormweaver",
				"-c",
				"fast",
				"-i",
				prevBackupDir .. "/backup_manifest"
			)

			w = pgm.primaryNode:make_worker("verification")
			w:calculate_database_checksums(backupDir .. ".checksum")
		end

		combine = { "-o", "datadirs/datadir_pr" }
		table.insert(combine, "backups/backup_0")

		for workloadIdx = 1, tonumber(args["repeat"]) do
			info("Restoring and verifying incremental #" .. tostring(workloadIdx))
			pgm:stop(1, 10)
			fs.delete_directory("datadirs/datadir_pr")

			table.insert(combine, "backups/backup_" .. tostring(workloadIdx))
			pgm:combinebackup(table.unpack(combine))
			pgm:start(1)

			backupChecksumFile = "backups/backup_" .. tostring(workloadIdx) .. ".checksum"
			checksumFile = "datadirs/datadir_pr/db.checksum"
			w = pgm.primaryNode:make_worker("reset")
			w:reset_metadata()
			w:discover_existing_schema()
			w:calculate_database_checksums(checksumFile)

			bp = BackgroundProcess.start("checksumdiff", "/usr/bin/diff", checksumFile, backupChecksumFile)
			if not bp:waitUntilExits() then
				error("Backup checksum verification failed with incremental #" .. tostring(workloadIdx))
			end
		end
	end)
end

File: scenarios/ci/incremental.py
# Incremental backup test: after each workload cycle take an incremental
# basebackup (PG 17+), then restore every backup chain prefix with
# pg_combinebackup and verify the database checksums match what was
# captured when the backup was taken.

import logging
import shutil
from pathlib import Path

from stormweaver import scenario

logger = logging.getLogger("scenario.incremental")


def add_arguments(parser):
    scenario.add_common_arguments(parser)


def main(args):
    opts = scenario.finalize(args)
    scenario.fresh_dir("backups", "archive")

    # archive=True also enables summarize_wal, required for incremental
    # basebackups
    with scenario.single_pg(opts, archive=True) as ctx:
        logger.info("taking base backup")
        ctx.pg.basebackup("backups/backup_0", extra_args=["-c", "fast"])

        for i in range(1, opts.repeat + 1):
            logger.info("cycle %d/%d: workload + incremental backup", i, opts.repeat)
            ctx.workload.run()

            prev = f"backups/backup_{i - 1}"
            backup = f"backups/backup_{i}"
            ctx.pg.basebackup(
                backup,
                extra_args=["-c", "fast"],
                incremental=f"{prev}/backup_manifest",
            )

            w = ctx.make_worker("verification")
            w.calculate_database_checksums(f"{backup}.checksum")
            logger.info("cycle %d/%d backed up", i, opts.repeat)

        chain = ["backups/backup_0"]
        for i in range(1, opts.repeat + 1):
            logger.info("Restoring and verifying incremental #%d", i)
            ctx.pg.stop()
            shutil.rmtree(ctx.datadir)

            chain.append(f"backups/backup_{i}")
            ctx.pg.combinebackup(chain, ctx.datadir)

            ctx.pg.start()
            if not ctx.pg.wait_ready():
                raise RuntimeError(f"restored server #{i} did not become ready")

            w = ctx.make_worker("reset")
            w.reset_metadata()
            w.discover_existing_schema()
            restored = Path(ctx.datadir) / "db.checksum"
            w.calculate_database_checksums(str(restored))

            scenario.compare_checksums(
                restored, f"backups/backup_{i}.checksum", f"incremental #{i}"
            )

    print("Incremental backup scenario completed successfully")
